zebra-state/src/request.rs

```diff
@@ -458,6 +458,15 @@ pub enum Request {
     /// * [`Response::Transaction(None)`](Response::Transaction) otherwise.
     Transaction(transaction::Hash),
 
+    /// Looks up a UTXO identified by the given [`OutPoint`](transparent::OutPoint),
+    /// returning `None` immediately if it is unknown.
+    ///
+    /// Checks verified blocks in the finalized chain and the _best_ non-finalized chain.
+    ///
+    /// This request is purely informational, there is no guarantee that
+    /// the UTXO remains unspent in the best chain.
```

Comment:
This isn't quite what we want - we should be checking for unspent `Utxo`s in the best chain. If the transaction output has been spent in the best chain, the mempool transaction is invalid and should be rejected.

It's ok to change the `BestChainUtxo` implementation to fix this, and rename it if you want.

We do a similar thing in `address_utxos()`, but the synchronisation required for multi-address, multi-utxo handling is much more complex:
https://github.com/ZcashFoundation/zebra/blob/cbc2e393caacaf82af3ac46603be344bd5cb2ea0/zebra-state/src/service/read/address/utxo.rs#L92

Since an `OutPoint` uniquely identifies a UTXO in a chain, we will only ever find one create and one spend per chain. So we don't need to duplicate the chain overlap and multi-address synchronization logic in `address_utxos()`. Any create in either finalized or non-finalized state makes the UTXO available, and any non-finalized spend makes it unavailable again. (The finalized state automatically deletes spent UTXOs from its index.)

So we'll need an `OutPoint` argument to the new read function. Then we can call `db.utxo()`,  `chain.created_utxo()`, and a new `chain.spent_utxo()` method directly, combine the results, and we're done.

Reply:
Updated in [`ad8d8ea`](https://github.com/ZcashFoundation/zebra/pull/5616/commits/ad8d8ea926c1415946f49b316c3c21616df6bb66).

I'm open to suggestions on naming.

Reply:
Happy to deal with naming later if we need to, the current name is clear enough!